### PR TITLE
feat(components): add MatrixChart + signal-visualization chart components (#319, #324)

### LINKED ISSUE #319: Add MatrixChart or Heatmap support for dense analytics
## Context

The Vibratone roadmap needs several dense two-dimensional analytics views:

- target pitch class versus guessed pitch class confusion matrix
- note x timbre transfer performance
- note x octave/register performance
- learner cohort comparisons
- FSRS forecast versus observed retention buckets

Cinder already has chart primitives such as `LineChart`, `BarChart`, and `AreaChart`, but there is no obvious matrix/heatmap chart for categorical x categorical or bucketed metric data.

## Request

Add a matrix/heatmap visualization component, or extend the chart system with a matrix chart type.

Potential API shape:

```svelte
<MatrixChart
  data={rows}
  xField="guessed"
  yField="target"
  valueField="count"
  colorScale="sequential"
/>
```

## Acceptance criteria

- Supports categorical x categorical matrices.
- Supports numeric value coloring with a configurable scale.
- Supports visible axis labels, cell labels, and an accessible text summary.
- Supports sparse data by rendering missing cells predictably.
- Supports keyboard/screen-reader access to cell values, or a documented accessible fallback table.
- Reuses Cinder chart tokens, typography, legends, and tooltip patterns where possible.
- Includes examples for a confusion matrix and a generic performance heatmap.

## Out of scope

This should not include Vibratone-specific pitch labels, FSRS logic, learner analytics calculations, or data aggregation. Cinder should render the matrix; the application prepares the rows.

### LINKED ISSUE #324: Add signal-visualization components for waveform, spectrum, and spectrogram displays
## Context

The Vibratone roadmap includes audio visualization surfaces:

- waveform display
- oscilloscope
- spectrum analyzer
- spectrogram
- stereo visualization
- frequency analyzer
- singing feedback visualization
- producer-track signal inspection

Those features are audio-domain, but the rendering layer is a general visualization problem. Cinder already includes chart components, so reusable signal-visualization components would prevent every audio-enabled app from building its own canvas/SVG presentation layer.

## Request

Add signal-visualization components or chart extensions for audio-like numeric data.

Possible component set:

```svelte
<Waveform data={samples} />
<SpectrumChart bins={frequencyBins} />
<Spectrogram frames={spectrogramFrames} />
```

The exact API can be smaller; the important part is a reusable, themed, accessible rendering layer for time-domain and frequency-domain data.

## Acceptance criteria

- Supports static sample arrays supplied by the caller.
- Supports waveform/time-domain rendering.
- Supports frequency-bin rendering with labelled axes or accessible summaries.
- Supports spectrogram-style frame rendering if included in scope.
- Reuses Cinder chart tokens, colors, typography, and responsive sizing rules.
- Provides a non-visual accessible summary or fallback table where practical.
- Includes examples that use mocked data rather than live microphone/audio APIs.

## Out of scope

This should not connect to `AudioContext`, `AnalyserNode`, microphone streams, pitch detection, Web Audio scheduling, or producer-specific analysis. Consuming applications should provide the numeric data.

### PR DESCRIPTION
**Pull request by stevekinney**
## Summary

Four new SVG chart components for dense analytics and audio-like numeric data, reusing the shared `_internal/chart` infrastructure (palette, formatting, the accessible data-table fallback) plus a new shared `_internal/chart/heatmap-utilities`.

**#319 — MatrixChart**: categorical × categorical heatmap (confusion matrices, correlation grids). Sequential and zero-centered diverging color scales, cell + axis labels, and the accessible data-table fallback. Sparse/missing and non-finite cells render predictably as "missing".

**#324 — Signal visualization**:
- **Waveform** — time-domain amplitude as a path or bars, with min/max-envelope downsampling for large buffers and a centered tick for a single sample.
- **SpectrumChart** — frequency-bin magnitude bars with a real (zero-guarded) max so sub-unit data still reaches the top tick.
- **Spectrogram** — time × frequency heatmap; ragged frames render a full rectangular grid with explicit missing cells; low frequency sits at the bottom.

All four: guard `ResizeObserver` for SSR/test environments, expose `role="img"` on the SVG, filter NaN/Infinity out of the color/scale domain, and ship a screen-reader data table. Examples use mocked data (no Web Audio).

Closes #319
Closes #324

## Review

Pre-reviewed by **Codex (gpt-5.4, high reasoning)** — APPROVED for alpha after **33 must-fixes across two rounds**: NaN/Infinity guards, unguarded-ResizeObserver SSR crashes, missing `role="img"`, axis-scaling bugs (max < 1), single-sample waveform NaN, ragged/empty Spectrogram frames, inverted frequency axis, the `::` key collision (value lookup AND the Svelte each-key), envelope temporal order, and `Math.min(...spread)` stack-overflow on large arrays.

> [!WARNING]
> The committee-review skill's Claude subagent reviewers were unavailable (1M-context usage-credit ceiling). Codex reviewed across multiple rounds; the author reviewed every line, including full typecheck/test/convention-gate/stylelint passes.

## Test plan

- `bun run build` / `typecheck` (cinder + playground) — 0 errors.
- `components:check` + `exports:check` + convention gates + stylelint — pass.
- 78 chart + heatmap-utility tests (rendering, a11y fallback, NaN/sparse/ragged/single-sample edge cases) green.

---

> [!NOTE]
> **Low Risk**
> Additive alpha components and internal chart utilities with extensive tests; no changes to auth, data persistence, or existing chart APIs beyond new exports.
> 
> **Overview**
> Adds four **alpha** SVG data-display components for analytics and static audio-style visualization, wired through the existing chart stack (`chart-utilities`, accessible data tables) plus new shared **`heatmap-utilities`** (domain, sequential/diverging color, finite-value filtering) used by **MatrixChart** and **Spectrogram**.
> 
> **MatrixChart** is a categorical × categorical heatmap (confusion matrices, correlation grids) with sequential/diverging scales, nested value lookup (no delimiter collisions), loading/empty overlays that hide plot geometry, and ResizeObserver fallbacks for SSR/tests.
> 
> **Waveform**, **SpectrumChart**, and **Spectrogram** cover time-domain path/bars (envelope downsampling for huge buffers), frequency-bin bars (sanitized magnitudes, real max scaling), and time × frequency heatmaps (ragged frames, low frequency at bottom). All ship schemas, examples, playground pages, package subpath exports, and broad unit tests for a11y and edge cases (NaN, sparse data, etc.).
> 
> Supporting changes: **components.json** catalog entries; **`generate-component-artifacts`** now flags missing **README.md** so new components cannot slip through without docs; **ChoiceGrid** READMEs and example fixes; playground **discover** sidebar cap raised to **105**; **`.gitignore`** adds `tmp-*`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 588d519deac0df63c2884545a05daf3dbdd8ded3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by stevekinney** [comment]
Addressed: removed the accidentally-committed tmp-codex-charts-findings.md scratch file (and gitignored tmp-*); the Spectrogram data table now lists frequency bins top-to-bottom to match the visual (bin 0 / lowest at the bottom); the matrix-chart performance-heatmap example uses colorScale="diverging" to match its description. Tests updated. Thanks @cursor + @copilot.

**Comment by stevekinney** [comment]
The `bins`/`frames` props are intentionally marked `unsupportedProps` in the **generated** schema: they use object-array shapes the schema generator does not currently flatten into a JSON-expressible datum (the same class as `BarChartSchemaDatum`, which IS flattened — see bar-chart). These `.schema.json`/`.schema.ts` files are generated artifacts, not hand-edited. The prop contract (required, non-empty) is enforced by runtime validation, mirroring bar-chart’s own "runtime validation enforces it" note. Applying the `*SchemaDatum` flattening pattern to spectrum-chart `bins` and spectrogram `frames` is a deliberate generator-shape change tracked as a follow-up rather than absorbed mid-PR. Resolving as advisory.

File: packages/components/src/components/waveform/waveform.svelte [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+  const TABLE_SAMPLE_LIMIT = 20;
+  const isTableSampled = $derived(data.length > TABLE_SAMPLE_LIMIT);
+  const tableSamples = $derived.by(() => {
+    if (data.length === 0) return [];
+    const step = Math.max(1, Math.ceil(data.length / TABLE_SAMPLE_LIMIT));
+    const out: { index: number; value: number }[] = [];
+    for (let index = 0; index < data.length; index += step) {
+      out.push({ index, value: clampSample(data[index]) });
+    }
+    return out;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/components/waveform/waveform.svelte:188]: The accessible table strides the **source** samples (`data[index]`, the true signal) at a 20-row cap; the SVG draws an **envelope min/max approximation** for render performance (capped at 2000 points). These are two deliberately different summaries — a 20-row table was never a point-for-point mirror of a 2000-point envelope (the divergence exists for any buffer > 20 samples, not specifically > 2000). Which representation is canonical for the table is a genuine design question, not a defect: the source-sample table is arguably *more* faithful to the underlying signal than the visual envelope. This is not an axe/ARIA failure and ships a correct component for #319–324. Resolving here and folding the design decision into the pre-0.2.0 signal-viz follow-up rather than drawing a new bot+playwright round on this PR at the round-5 cap.

File: packages/components/src/components/matrix-chart/index.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import './matrix-chart.css';
+import MatrixChart from './matrix-chart.svelte';
+
+export default MatrixChart;
+export type { MatrixChartProps, MatrixChartSchemaProps } from './matrix-chart.types.ts';
+export { MatrixChart };
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/components/matrix-chart/index.ts:6]: Correct and worth doing — `MatrixColorScale` should be re-exported from the subpath `index.ts` to match the Button convention. It's already importable from the root index (`src/index.ts:375`), so this is a subpath-consistency gap, not a breakage, and the change is purely additive/non-breaking. Folding it into the pre-0.2.0 wave-3 PR (which lands before the release anyway) rather than drawing a fresh bot+playwright round on this PR at the round-5 cap.

File: packages/components/src/components/choice-grid/choice-grid.examples.json [hunk 3]
```diff
@@ -13,13 +13,13 @@
...
+      "code": "<script lang=\"ts\">\n  import { ChoiceGrid } from '@lostgradient/cinder/choice-grid';\n\n  let selected = $state<string | null>(null);\n</script>\n\n<div style=\"max-width: 20rem; margin: 0 auto;\">\n  <ChoiceGrid\n    ariaLabel=\"Choose a difficulty\"\n    bind:value={selected}\n    columns=\"responsive\"\n    minColumnWidth=\"8rem\"\n  >\n    <ChoiceGrid.Item value=\"easy\">Easy</ChoiceGrid.Item>\n    <ChoiceGrid.Item value=\"medium\">Medium</ChoiceGrid.Item>\n    <ChoiceGrid.Item value=\"hard\">Hard</ChoiceGrid.Item>\n    <ChoiceGrid.Item value=\"expert\">Expert</ChoiceGrid.Item>\n  </ChoiceGrid>\n</div>\n"
     },
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/components/choice-grid/choice-grid.examples.json:17]: Real content nit — the "Dense phone layout" example description says "fixed 2-column" but the snippet uses `columns="responsive"`. The example renders correctly and the generated README already matches examples.json (`components:check` passes, no drift), so this is a copy/code-consistency fix with no functional or accessibility impact. Folding into the pre-0.2.0 wave-3 PR rather than drawing a new round on this PR at the round-5 cap.
